lib/DxcSupport/HLSLOptions.cpp

```diff
@@ -69,6 +69,116 @@ const OptTable * hlsl::options::getHlslOptTable() {
   return g_HlslOptTable;
 }
 
+static std::vector<ArgPair> ComputeArgPairsFromArgList(const llvm::opt::InputArgList &argList) {
+  std::vector<ArgPair> ret;
+
+  for (llvm::opt::Arg *arg : argList) {
+    if (arg->getOption().matches(OPT_INPUT))
+      continue;
+
+    llvm::StringRef name = arg->getOption().getName();
+    llvm::StringRef value;
+
+    llvm::SmallString<64> argumentStorage;
+
+    // If this is a positional argument, set the name to ""
+    // explicitly.
+    if (arg->getOption().getKind() == llvm::opt::Option::InputClass) {
+      assert(arg->getNumValues() == 0);
+      name = "";
+      value = arg->getValue();
+    }
+    // This is a flag without value, like -Zi, or -Od
+    else if (arg->getOption().getKind() == llvm::opt::Option::FlagClass) {
+      assert(arg->getNumValues() == 0);
+    }
+    else if (arg->getNumValues() == 1) {
+      value = arg->getValue();
+      // If the argument must be merged (eg. -Wx, where W is the option and x is
+      // the value), merge them right now.
+      if (arg->getOption().getKind() == llvm::opt::Option::JoinedClass) {
+        argumentStorage.append(name);
+        argumentStorage.append(value);
+
+        name = argumentStorage;
+        value = "";
+      }
+    }
+    else if (arg->getOption().getKind() == Option::CommaJoinedClass) {
+      //
+      // CommaJoinedClass takes the following form:
+      //  -MyArg=Value0,Value1,Value2,Value3
+      //
+      // Much like JoinedClass, it gets folded into the name.
+      //
+      //  { "MyArg=Value0,Value1,Value2,Value3", "" }
+      argumentStorage += name;
+      for (unsigned i = 0; i < arg->getNumValues(); i++) {
+        const char *valuePtr = arg->getValue(i);
+        if (i > 0)
+          argumentStorage += ",";
+        argumentStorage += valuePtr;
+      }
+      name = argumentStorage;
+      value = "";
+    }
+
+    // This type of arg takes the following form:
+    //  -arg Value0 Value1 Value2
+    //
+    // To preserve the arg correctly, we must represent it as the following:
+    //   { "arg", "" }
+    //   { "",    "Value0" }
+    //   { "",    "Value1" }
+    //   { "",    "Value2" }
+    // 
+    // ..Which is not ideal, since the whole point of arg pair is to easily break down what arg is part of what.
+    // 
+    // Alternative is to try this:
+    //   { "arg", "Value0 Value1 Value2" }
+    // 
+    // But this causes problem when recompiling
```

Comment:
> "But this causes problem when recompiling"

What does this mean?  Is the problem that it doesn't round-trip properly - ie you can't trivially reconstruct a command-line when it's stored in this alternate form?

Reply:
Added an explanation for how it actually failed.